### ISSUE TITLE
Postfix config misses smtp_* parameters and includes several default settings

### ISSUE THREAD

**Issue by gsauthof** [mozilla/ssl-config-generator#114]
Take this generated config: https://ssl-config.mozilla.org/#server=postfix&version=3.5.6&config=intermediate&openssl=1.1.1g&guideline=5.6

The following values are already Postfix defaults:

```
smtpd_tls_mandatory_ciphers = medium
tls_preempt_cipherlist = no
```

I've checked the Postfix documentation and `postconf -d` output on Fedora 32.

Then there are some settings for when doing mandatory TLS, i.e.

```
smtpd_tls_mandatory_protocols = !SSLv2, !SSLv3, !TLSv1, !TLSv1.1
smtpd_tls_mandatory_ciphers = medium
```

although mandatory TLS isn't configured - cf. the `smtpd_tls_security_level = may` line which configures opportunistic TLS. Thus, those options aren't effective here.

On the other hand, the sister options for opportunistic TLS are missing, i.e. this one is missing:

```
smtpd_tls_protocols = !SSLv2, !SSLv3, !TLSv1, !TLSv1.1
```

Also, the generated config only includes `smtpd_*` options and no `smtp_*` variants. Note that opportunistic TLS also makes sense when the mail server is sending mail, i.e. e.g. when SMTP relaying mail to a destination server. Thus, the config is missing at least the following smtp options:

```
smtp_tls_security_level = may
smtp_tls_protocols  = !SSLv2, !SSLv3, !TLSv1, !TLSv1.1
```

**Comment by DG1000** [comment]
I agree. I just updated my configuration and was wondering why the `smtp_*` variants are missing.

**Comment by polarathene** [comment]
<details>
<summary>Regarding cipher selection</summary>

I would imagine that `smtp_*` aren't set as delivery via port 25:
- Is expected to support plain-text if STARTTLS is unable to establish a secure connection.
- Nor is a secure connection to the MTA necessarily secure throughout the full transit to the recipient as there may be other hops involved afterwards out of your influence.

If you choose to restrict the ciphers that could be used, the fallback thus is to continue the connection and deliver without any encryption.

For connections to a relay host like SendGrid you may prefer to set the security level to `encrypt` instead of `my` to avoid plain-text, then it is up to SendGrid from then on which will probably be much more relaxed. Ideally you'd also relay over implicit TLS (port 465) instead of opportunistic (port 587).

</details>

---

- [`smtp_tls_security_level = may`](https://www.postfix.org/postconf.5.html#smtp_tls_security_level) 👍 
- [`smtp_tls_protocols = >=TLSv1.2`](https://www.postfix.org/postconf.5.html#smtp_tls_protocols) (_the equivalent syntax since Postfix 3.6, the default for `smtp_` here is `>=TLSv1`_)

**Note:**
- If the MTA you're delivering to for some reason still doesn't support TLS 1.2, you're preferring to deliver without any encryption at all vs a cipher that may still provide some degree of security.
- There are [still some old systems in 2024 in use that require TLS <1.2](https://github.com/docker-mailserver/docker-mailserver/pull/2945#issuecomment-1949663267).

---

> The following values are already Postfix defaults:
> 
> ```
> smtpd_tls_mandatory_ciphers = medium
> tls_preempt_cipherlist = no
> ```

It's ok to be explicit with security settings that are still the defaults. A user may be applying the suggested config to an already configured instance where these may have been configured differently by mistake or intentionally.

IIRC:
- [`tls_preempt_cipherlist = yes`](https://www.postfix.org/postconf.5.html#tls_preempt_cipherlist) is fine for port 25 where you're cipher support may be more broad for compatibility. The server provides preference order and the client must choose the first one it is compatible with.
- `tls_preempt_cipherlist = no` can be acceptable on submission(s) ports 587 and 465 where you may enforce TLS and restrict the cipher list for authenticating clients to those that are only forward secret compatible (**EDIT:** May have been AEAD specific.. it's been a while since I looked into this_). Letting the client choose is better here.

While mandatory ciphers is set to `medium`, some may adjust this to `high` to alter the `high` cipher list if they customize the cipher list (against the advice of Postfix). Which can be useful when you use a security level of `encrypt`.

The config only focuses on `main.cf`, and not `master.cf` where the submission(s) ports will apply their overrides for the additional security expectations. [`smtpd_tls_auth_only = yes`](https://www.postfix.org/postconf.5.html#smtpd_tls_security_level) presently used in the generator shouldn't be relevant if port 25 doesn't offer SASL auth (_which it shouldn't, but you do see this in the wild_), it's implicit when using a security level of `encrypt`, by default SASL auth is disabled as per [`smtpd_sasl_auth_enable`](https://www.postfix.org/postconf.5.html#smtpd_sasl_auth_enable).

[`smtpd_tls_exclude_ciphers`](https://www.postfix.org/postconf.5.html#smtpd_tls_exclude_ciphers) can be relevant if you're comfortable dismissing the advice to avoid configuring it from Postfix docs. [DMS sets this](https://github.com/docker-mailserver/docker-mailserver/blob/a815bf5ab44388baecb597a0b83f71872a2d34bc/target/postfix/main.cf#L47C1-L47C27) (_see `git blame` history for reference, [this explains SEED / CAMELLIA](https://github.com/docker-mailserver/docker-mailserver/pull/1802#issuecomment-778033195), and [this for SHA1](https://github.com/docker-mailserver/docker-mailserver/pull/2945)_).

**Comment by gstrauss** [comment]
I filed PR #278 for this issue.

@polarathene wrote:
> * If the MTA you're delivering to for some reason still doesn't support TLS 1.2, you're preferring to deliver without any encryption at all vs a cipher that may still provide some degree of security.

Note: `smtp_tls_security_level` defaults to none, so prior behavior is no TLS when sending email.  Therefore, it is ok for PR #278 to set cipher restrictions for `smtp_tls_*` with `smtp_tls_security_level=may`.  If the restrictions are not met, TLS is not used, and the fallback is clear-text, but that is no different than the current behavior of effectively `smtp_tls_security_level=none`.

> * There are [still some old systems in 2024 in use that require TLS <1.2](https://github.com/docker-mailserver/docker-mailserver/pull/2945#issuecomment-1949663267).

Those deficient and poorly maintained systems -- which are ancient and reached end-of-life a long time ago -- will fall back to using clear-text, or the `smtp_tls_*` directives can be adjusted to be from the Old configuration.

> * [`tls_preempt_cipherlist = yes`](https://www.postfix.org/postconf.5.html#tls_preempt_cipherlist) is fine for port 25 where your cipher support may be more broad for compatibility. The server provides preference order and the client must choose the first one it is compatible with.
> 
> * `tls_preempt_cipherlist = no` can be acceptable on submission(s) ports 587 and 465 where you may enforce TLS and restrict the cipher list for authenticating clients to those that are only forward secret compatible (_**EDIT:** May have been AEAD specific.. it's been a while since I looked into this_). Letting the client choose is better here.

@polarathene: I think you have a misunderstanding of how and why server cipher selection works when using server preference, and its implications.  While the rest of what you posted makes sense, I think your notes about tls_preempt_cipherlist (which I quoted above) are misleading.

> While mandatory ciphers is set to `medium`, some may adjust this to `high` to alter the `high` cipher list if they customize the cipher list (against the advice of Postfix). Which can be useful when you use a security level of `encrypt`.

The defaults for `medium` and `high` built into Postfix appear to be circa 2015 and are dated, to say the least.  ssl-config-generator recommends stricter security settings for today, which is almost 10 years later than 2015.

**Comment by gstrauss** [comment]
I filed PR #278 for this issue.

@polarathene wrote:
> * If the MTA you're delivering to for some reason still doesn't support TLS 1.2, you're preferring to deliver without any encryption at all vs a cipher that may still provide some degree of security.

Note: `smtp_tls_security_level` defaults to none, so prior behavior is no TLS when sending email.  Therefore, it is ok for PR #278 to set cipher restrictions for `smtp_tls_*` with `smtp_tls_security_level=may`.  If the restrictions are not met, TLS is not used, and the fallback is clear-text, but that is no different than the current behavior of effectively `smtp_tls_security_level=none`.

> * There are [still some old systems in 2024 in use that require TLS <1.2](https://github.com/docker-mailserver/docker-mailserver/pull/2945#issuecomment-1949663267).

Those deficient and poorly maintained systems -- which are ancient and reached end-of-life a long time ago -- will fall back to using clear-text, or the `smtp_tls_*` directives can be adjusted to be from the Old configuration.

> * [`tls_preempt_cipherlist = yes`](https://www.postfix.org/postconf.5.html#tls_preempt_cipherlist) is fine for port 25 where your cipher support may be more broad for compatibility. The server provides preference order and the client must choose the first one it is compatible with.
> 
> * `tls_preempt_cipherlist = no` can be acceptable on submission(s) ports 587 and 465 where you may enforce TLS and restrict the cipher list for authenticating clients to those that are only forward secret compatible (_**EDIT:** May have been AEAD specific.. it's been a while since I looked into this_). Letting the client choose is better here.

@polarathene: I think you have a misunderstanding of how and why server cipher selection works when using server preference, and its implications.  While the rest of what you posted makes sense, I think your notes about tls_preempt_cipherlist (which I quoted above) are misleading.
(Edit: e.g. see https://github.com/mozilla/ssl-config-generator/issues/157#issuecomment-2408905628)

> While mandatory ciphers is set to `medium`, some may adjust this to `high` to alter the `high` cipher list if they customize the cipher list (against the advice of Postfix). Which can be useful when you use a security level of `encrypt`.

The defaults for `medium` and `high` built into Postfix appear to be circa 2015 and are dated, to say the least.  ssl-config-generator recommends stricter security settings for today, which is almost 10 years later than 2015.

**Comment by polarathene** [comment]
> If the restrictions are not met, TLS is not used, and the fallback is clear-text, but that is no different than the current behavior of effectively `smtp_tls_security_level=none`.

From the context of my response, I'm going to assume that was in regards to the other line I was responding to related to that change to `may`, where the min TLS level was raised to 1.2.

There is [very little valid cipher suites for TLS below 1.2](https://github.com/docker-mailserver/docker-mailserver/issues/2679) these days, but they technically still offer some security over none at all.

---

> Those deficient and poorly maintained systems -- which are ancient and reached end-of-life a long time ago -- will fall back to using clear-text, or the `smtp_tls_*` directives can be adjusted to be from the Old configuration.

Yes I agree. It was just additional context, given those environments can still depend on the security advice of a project like this it is sometimes worth having the awareness of what those environments are like in 2024. Not that they need to be supported.

---

> @polarathene: I think you have a misunderstanding of how and why server cipher selection works when using server preference, and its implications. While the rest of what you posted makes sense, I think your notes about `tls_preempt_cipherlist` (which I quoted above) are misleading.

I disagree with you here:
- Port 25 should allow clear-text delivery, and may offer broader cipher suite selection to avoid that. Enforcing server selection is valid when you've defined the order properly (as DMS does).
- Submissions ports (587 + 465) is a different client type. There's trust through credentials and they should be more modern, TLS can also be enforced. The server is thus more likely to restrict to AEAD ciphers only, where the client can be free to choose which one is most compatible for them. Very similar to how TLS 1.3 works.

Where am I being misleading? What am I misunderstanding? This was a topic I looked into years ago yes, but I'm still fairly confident with that differentiation of ports.

If you offer a set of ciphers where there is no concerns with choice, it's better to let the client choose. Otherwise the server can enforce priority with the first compatible cipher suite in it's list.

> (Edit: e.g. see https://github.com/mozilla/ssl-config-generator/issues/157#issuecomment-2408905628)

You linked to an Oct 2024 response of yours to a Jan 2022 comment of mine on the topic. Compared to the Feb 2024 comment of mine here. That's a 2 year difference. Could you please add more context to the point of your reference?

---

> > While mandatory ciphers is set to `medium`, some may adjust this to `high` to alter the `high` cipher list if they customize the cipher list (against the advice of Postfix). Which can be useful when you use a security level of `encrypt`.
> 
> The defaults for `medium` and `high` built into Postfix appear to be circa 2015 and are dated, to say the least. ssl-config-generator recommends stricter security settings for today, which is almost 10 years later than 2015.

As someone who maintains a mail server project (one of the most starred ones on Github IIRC), I'm not sure if you were paying attention to what I had said about `medium` and `high` cipher list selection.

Please look at this test case of mine:
- `high` (port 587/465) => [RSA `intermediate` vs `modern`](https://github.com/docker-mailserver/docker-mailserver/blob/edfecbceb163eda98124fdb10dd37bfe6fa0f46d/test/tests/parallel/set2/tls_cipherlists.bats#L208-L210) (_these are not intended to match the equivalent `ssl-config-generator` lists btw_)
- `medium` (port 25) => [RSA](https://github.com/docker-mailserver/docker-mailserver/blob/edfecbceb163eda98124fdb10dd37bfe6fa0f46d/test/tests/parallel/set2/tls_cipherlists.bats#L221-L223)

For the `high` list, [it is configured during startup to apply the intermediate or modern cipher suite lists](https://github.com/docker-mailserver/docker-mailserver/blob/edfecbceb163eda98124fdb10dd37bfe6fa0f46d/target/scripts/helpers/ssl.sh#L92-L97).

Other settings are configured elsewhere, [along with some defaults of ours](https://github.com/docker-mailserver/docker-mailserver/blob/edfecbceb163eda98124fdb10dd37bfe6fa0f46d/target/postfix/main.cf#L42-L47), such as when the `high` list will be used. You'd also notice that while we don't overwrite the `medium` list, we do restrict it via protocol and exclusions (as per last link).

With those changes the `medium` list as per the test suite link for it, is quite secure in it's own right. They're not all AEAD though, so server selection is used. When I did these contributions, at the time I was fairly new to all of it (late 2020), but I did invest a significant amount of time becoming familiar with the ciphersuites and wrote up about 20 pages of revised/condensed notes on the topic IIRC.

So to clarify. DMS uses both `medium` and `high` lists, with a customized `high` list, along with raising the min TLS protocol version to 1.2 and using and excludes list to trim `medium` cipher suites down to a set I deemed secure. `medium` was used for SMTP port 25, while the submission(s) ports used the mandatory cipher list set to `high`.

**Comment by gstrauss** [comment]
> Where am I being misleading? What am I misunderstanding?

>> [tls_preempt_cipherlist = yes](https://www.postfix.org/postconf.5.html#tls_preempt_cipherlist) is fine for port 25 where your cipher support may be more broad for compatibility. The server provides preference order and the client must choose the first one it is compatible with.

Specifically, I think that is incorrect or unclear.  Postfix `tls_preempt_cipherlist = yes` means server chooses, not the client.  I am interpreting this in the context where Postfix is the server, and is receiving connections on port 25.

Generally speaking, my post is not written and directed at you, but rather is written for others reading your post and then mine.  Please do not take that personally.  Let me try to explain briefly:

> If you offer a set of ciphers where there is no concerns with choice, it's better to let the client choose.

"Offer a set of ciphers where this is no concerns with choice" requires a certain amount of knowledge that you have, but the average reader might not have, as the average reader is not an expert in cryptography and did not write a paper about DHEater.

Many readers will not know AEAD or what it means to PFS, or what PFS is and why PFS matters.

Given the *current* recommendations on *this site* (ssl-config-generator), which are not restricted to PFS-only ciphers for Intermediate or Old configs, it is important to me that I highlight it **IS NOT** better to let the client choose, since the ciphers *currently* recommended on *this site* for Intermediate and Old configurations are not PFS-only (so attacks might try to trigger a downgrade to a non-PFS cipher), hence my use of the adjective "misleading".  You may substitute that with "confusing" or "not clear", none of which are positive when trying to communicate advice for how to reasonably configure TLS for an application to be reasonably secure.

As I said already, this site is ssl-config-generator, and is not docker-mailserver, and does not have docker-mailserver configs or docker-mailserver TLS customizations.  Again, my post is for others reading this, not for you as you have clearly spent time configuring and tuning TLS for docker-mailserver and I recognize that as a good thing for docker-mailserver.

---

My general statement about fallback to clear-text was intended to describe that the changes I made in #278 are no worse a situation than the current configuration with the default settings implied for Postfix, which is why I am side-stepping whether allowing fallback to TLSv1 would be better than TLSv1.2 or fallback to clear-text.  For the Intermediate configuration, #278 uses TLSv1.2 and doing so avoids certain classes of downgrade attacks which might cause a weaker cipher to be chosen.  Of course, fallback to clear-text is insecure if fallback to clear-text occurs.  I do not wish to argue which might be better.  I made a configuration change to the recommendations and the overall recommendations (enabling `smtp_tls_security_level=may`) is an improvement over the prior recommendations.

Please understand that I am not trying to debate you here.  I made this post as a courtesy to attempt to answer your questions: "Where am I being misleading? What am I misunderstanding?"

Cheers.

**Comment by polarathene** [comment]
**TL;DR:**
- We differ in opinion a bit due to perspective of the reader, but we're roughly on the same page AFAIK 👍
- There was a minor regression with #278 (_not with outbound connections, but inbound_), nothing serious IMO :)
- Nothing taken personally, I appreciate you taking the time to provide feedback.

No expectation for a further response to this, like you I'm not here to argue just inform. I care about being correct not right, perfectly happy when I'm wrong so long as I can learn from it.

---

To the readers landing here
- If the generator output isn't satisfactory for you, please don't just blindly take my advice or copy whatever you see over at `docker-mailserver` config wise - especially if you don't understand the jargon used. I have experience and know enough on the topic to engage in this discussions at a more technical level, but I don't consider myself an expert / specialist.
- If you want to ask a question that is difficult to research an answer for (_you've tried to find an answer for over an hour_), feel free to ping me if it seems like I might be able to clear something up.

---

> Specifically, I think that is incorrect or unclear. Postfix `tls_preempt_cipherlist = yes` means server chooses, not the client. I am interpreting this in the context where Postfix is the server, and is receiving connections on port 25.

Yes we are on the same page there with server and port. IIRC the client sends what ciphers it is compatible with and the server chooses the the first match on it's list.

If the client restricts the list of compatible ciphers offered, then they are able to manipulate the selection a bit, which is generally where it's relevant attack wise? Port 25 though, should always support delivery without TLS, so that is less relevant.

For port 587 and 465 you can make it mandatory to use TLS with `=encrypt` instead of `=may`, and in this context it makes sense to restrict to AEAD cipher suites only where order of the list has no effect on selection.

---

> a certain amount of knowledge that you have, but the average reader might not have

> Many readers will not know AEAD or what it means to PFS, or what PFS is and why PFS matters.

My response is in the context of the issue and project. Any reader here trying to learn technical details while lacking the context of what is being discussed should not be blindly applying it? That's the whole point of the official config generator for users.

If this discussion were to take place on some platform like reddit, I'd agree with you more but I'd still side with the notion of not trusting information from unofficial sources blindly (_granted, in this project I am marked as "contributor"_). My responses tend to be verbose enough as-is, but when I feel there's value to add more information to what is being discussed I tend to add it, especially if it's not as easy to lookup.

> it is important to me that I highlight it **IS NOT** better to let the client choose, since the ciphers _currently_ recommended on _this site_ for Intermediate and Old configurations are not PFS-only (so attacks might try to trigger a downgrade to a non-PFS cipher), hence my use of the adjective "misleading".

That's fair. Although if a reader were to follow the advice from my comment, at the end I do mention and reference exclusion setting along with the earlier TLS 1.2 min protocol version, which as my follow-up comment shared the `medium` test case which last I recall is only cipher suites that are [PFS](https://en.wikipedia.org/wiki/Forward_secrecy).

> You may substitute that with "confusing" or "not clear", none of which are positive when trying to communicate advice for how to reasonably configure TLS for an application to be reasonably secure.

Yes, because this issue is for discussing change and security, where actual endorsed advice is through merged PRs and the resulting tools that leverage that to advise those without the expertise.

Readers that land here may be curious and may want more information to understand something better. They can ask questions or do further research if they like.

---

> As I said already, this site is ssl-config-generator, and is not docker-mailserver, and does not have docker-mailserver configs or docker-mailserver TLS customizations.

Yes, I use DMS as a reference for additional context with real-world deployment that differs a bit in opinion. We're also discussing mail servers in this issue which have been known historically to lag behind quite a bit vs web servers (_or at least did quite a bit when I last looked into it, both on server and client sides_), I consider that relevant context.

Someone may use the config advice generated from this project for example with software they are using that already provides some non-default configuration. I raised that concern earlier as well for why it's ok to be more explicit with settings that would otherwise be omitted since they are the default.

Someone that tries to mix the two config sources may be taking a risk if they're not careful. Same applies if they were to nitpick advice on specific settings above without taking into account the rest of the context.

> My general statement about fallback to clear-text was intended to describe that the changes I made in #278 are no worse a situation than the current configuration with the default settings implied for Postfix, which is why I am side-stepping whether allowing fallback to TLSv1 would be better than TLSv1.2 or fallback to clear-text.

Yeah I understood that too. There was a regression you may have overlooked from the change AFAIK, I'll clarify that.

The generator suggests `smtpd_tls_security_level = may` which is appropriate for port 25, but port 587/465 should technically favor `smtpd_tls_security_level = encrypt` (_needs to be done in `master.cf`, not `main.cf`_). As `=may` for outbound port 25 is not a concern with your PR, the existing inbound port 25 also has `=may` set, which will regress cipher suite availability there, but arguably it's not a major concern given clients of the submission(s) ports are more akin to web, only old phones, scanners and other hardware will lose access (_as pointed out by my reference to known affected users in 2024_).

---

> I do not wish to argue which might be better.

No worries, I'm not here to argue either. Just ensuring any misunderstandings on either end are cleared up 👍 

> (enabling `smtp_tls_security_level=may`) is an improvement over the prior recommendations.

Agreed. DMS has had that for some time, I'm a bit surprised that Postfix defaults to `none` for outbound SMTP connections 🤔 

---

> Please understand that I am not trying to debate you here. I made this post as a courtesy to attempt to answer your questions: "Where am I being misleading? What am I misunderstanding?"

> Please do not take that personally.

Oh no worries! I'm usually the one to raise these disclaimers as my communication is often taken the wrong way 😓 😝 

My questions are purely that, no hidden intent or desire to be defensive. I only care about us correcting any misinformation/misunderstandings for the benefit of everyone :)
